src/authentication_manager.rs

```diff
@@ -43,15 +43,55 @@ impl AuthenticationManager {
     #[tracing::instrument]
     pub async fn new() -> Result<Self, Error> {
         tracing::debug!("Initializing gcp_auth");
-        if let Some(service_account) = CustomServiceAccount::from_env()? {
-            return Ok(service_account.into());
+        let client = types::client();
+        if let Some(service_account_creds) = CredentialSource::from_env().await? {
+            tracing::debug!("Using GOOGLE_APPLICATION_CREDENTIALS env");
+
+            let service_account: Box<dyn ServiceAccount> = match service_account_creds {
+                CredentialSource::ServiceAccount(creds) => {
+                    let service_account = CustomServiceAccount::new(creds)?;
+                    Box::new(service_account)
+                }
+                CredentialSource::AuthorizedUser(creds) => {
+                    let service_account =
+                        ConfigDefaultCredentials::from_user_credentials(creds, &client).await?;
+                    Box::new(service_account)
+                }
+            };
+
+            return Ok(Self {
+                service_account,
+                client,
+                refresh_mutex: Mutex::new(()),
+            });
         }
 
-        let client = types::client();
-        let default_user_error = match ConfigDefaultCredentials::new(&client).await {
-            Ok(service_account) => {
+        let default_user_error = match CredentialSource::from_default_credentials().await {
+            Ok(service_account_creds) => {
                 tracing::debug!("Using ConfigDefaultCredentials");
-                return Ok(Self::build(client, service_account));
+
+                let service_account: Result<Box<dyn ServiceAccount>, Error> =
+                    match service_account_creds {
+                        CredentialSource::AuthorizedUser(creds) => {
+                            ConfigDefaultCredentials::from_user_credentials(creds, &client)
+                                .await
+                                .map(|creds| Box::new(creds) as _)
```

Comment:
Compiler can't infer `Result<Box<Struct>, _>` into `Result<Box<dyn Trait>, _>` which is why this cast is needed.